File: polyglot/go/evolve/evolve.go
// Package evolve implements a simple GP engine using arithmetic expression trees.
package evolve

import (
	"fmt"
	"math/rand"
	"strings"
)

// Expr is an arithmetic expression tree node.
type Expr interface {
	Eval() int64
	Depth() int
	Size() int
	Clone() Expr
	String() string
}

// Num is a constant integer.
type Num struct{ Val int64 }

func (n *Num) Eval() int64    { return n.Val }
func (n *Num) Depth() int     { return 1 }
func (n *Num) Size() int      { return 1 }
func (n *Num) Clone() Expr    { return &Num{Val: n.Val} }
func (n *Num) String() string { return fmt.Sprintf("%d", n.Val) }

// BinOp is a binary operation.
type BinOp struct {
	Op          string
	Left, Right Expr
}

func (b *BinOp) Eval() int64 {
	l, r := b.Left.Eval(), b.Right.Eval()
	switch b.Op {
	case "+":
		return l + r
	case "-":
		return l - r
	case "*":
		return l * r
	case "mod":
		if r == 0 {
			return 0
		}
		return l % r
	}
	return 0
}

func (b *BinOp) Depth() int {
	ld, rd := b.Left.Depth(), b.Right.Depth()
	if ld > rd {
		return ld + 1
	}
	return rd + 1
}

func (b *BinOp) Size() int  { return 1 + b.Left.Size() + b.Right.Size() }
func (b *BinOp) Clone() Expr {
	return &BinOp{Op: b.Op, Left: b.Left.Clone(), Right: b.Right.Clone()}
}
func (b *BinOp) String() string {
	return fmt.Sprintf("(%s %s %s)", b.Op, b.Left.String(), b.Right.String())
}

var ops = []string{"+", "-", "*", "mod"}

// RandomExpr generates a random expression tree up to the given depth.
func RandomExpr(rng *rand.Rand, maxDepth int) Expr {
	if maxDepth <= 1 || rng.Intn(3) == 0 {
		return &Num{Val: int64(rng.Intn(101))}
	}
	return &BinOp{
		Op:    ops[rng.Intn(len(ops))],
		Left:  RandomExpr(rng, maxDepth-1),
		Right: RandomExpr(rng, maxDepth-1),
	}
}

// Mutate returns a mutated copy of the expression.
func Mutate(rng *rand.Rand, expr Expr) Expr {
	clone := expr.Clone()
	switch rng.Intn(3) {
	case 0: // Change a constant
		mutateNum(rng, clone)
	case 1: // Change an operator
		mutateOp(rng, clone)
	case 2: // Replace a subtree
		return replaceRandom(rng, clone, RandomExpr(rng, 2))
	}
	return clone
}

func mutateNum(rng *rand.Rand, expr Expr) {
	switch e := expr.(type) {
	case *Num:
		e.Val += int64(rng.Intn(21)) - 10
	case *BinOp:
		if rng.Intn(2) == 0 {
			mutateNum(rng, e.Left)
		} else {
			mutateNum(rng, e.Right)
		}
	}
}

func mutateOp(rng *rand.Rand, expr Expr) {
	if b, ok := expr.(*BinOp); ok {
		if rng.Intn(2) == 0 && b.Left.Depth() > 1 {
			mutateOp(rng, b.Left)
		} else if b.Right.Depth() > 1 {
			mutateOp(rng, b.Right)
		} else {
			b.Op = ops[rng.Intn(len(ops))]
		}
	}
}

func replaceRandom(rng *rand.Rand, expr Expr, replacement Expr) Expr {
	if rng.Intn(3) == 0 {
		return replacement
	}
	if b, ok := expr.(*BinOp); ok {
		if rng.Intn(2) == 0 {
			b.Left = replaceRandom(rng, b.Left, replacement)
		} else {
			b.Right = replaceRandom(rng, b.Right, replacement)
		}
	}
	return expr
}

// Crossover combines two expression trees.
func Crossover(rng *rand.Rand, a, b Expr) Expr {
	ac := a.Clone()
	bc := b.Clone()
	// Pick a random subtree from b and insert into a.
	donor := pickRandom(rng, bc)
	return replaceRandom(rng, ac, donor)
}

func pickRandom(rng *rand.Rand, expr Expr) Expr {
	if b, ok := expr.(*BinOp); ok && rng.Intn(3) > 0 {
		if rng.Intn(2) == 0 {
			return pickRandom(rng, b.Left)
		}
		return pickRandom(rng, b.Right)
	}
	return expr.Clone()
}

// Candidate is a GP individual.
type Candidate struct {
	Expr    Expr
	Fitness float64
}

// Score evaluates a candidate against a target output value.
func Score(expr Expr, targetVal int64) float64 {
	result := expr.Eval()
	if result == targetVal {
		return 1000.0 - float64(expr.Size())*10.0
	}
	diff := result - targetVal
	if diff < 0 {
		diff = -diff
	}
	if diff > 1000 {
		return 0.0
	}
	return 1.0 / (1.0 + float64(diff))
}

// RunEvolution runs the GP engine against a target value.
// Returns the best expression and whether it found an exact solution.
func RunEvolution(targetVal int64, maxGens int, seed int64) (Expr, bool) {
	rng := rand.New(rand.NewSource(seed))
	popSize := 30
	elites := 5

	// Initialize population.
	pop := make([]Candidate, popSize)
	for i := range pop {
		pop[i] = Candidate{Expr: RandomExpr(rng, 4)}
	}

	for gen := 0; gen < maxGens; gen++ {
		// Evaluate.
		for i := range pop {
			pop[i].Fitness = Score(pop[i].Expr, targetVal)
		}

		// Sort by fitness descending.
		sortPop(pop)

		// Check for winner.
		if pop[0].Fitness >= 900.0 {
			return pop[0].Expr, true
		}

		// Next generation.
		next := make([]Candidate, 0, popSize)
		for i := 0; i < elites && i < len(pop); i++ {
			next = append(next, Candidate{Expr: pop[i].Expr.Clone()})
		}
		for len(next) < popSize {
			parent := tournamentSelect(rng, pop)
			var child Expr
			if rng.Intn(5) == 0 {
				other := tournamentSelect(rng, pop)
				child = Crossover(rng, parent.Expr, other.Expr)
			} else {
				child = Mutate(rng, parent.Expr)
			}
			// Limit tree size.
			if child.Size() > 20 {
				child = RandomExpr(rng, 3)
			}
			next = append(next, Candidate{Expr: child})
		}
		pop = next
	}

	// Return best found.
	for i := range pop {
		pop[i].Fitness = Score(pop[i].Expr, targetVal)
	}
	sortPop(pop)
	return pop[0].Expr, pop[0].Fitness >= 900.0
}

func tournamentSelect(rng *rand.Rand, pop []Candidate) Candidate {
	best := pop[rng.Intn(len(pop))]
	for i := 0; i < 2; i++ {
		c := pop[rng.Intn(len(pop))]
		if c.Fitness > best.Fitness {
			best = c
		}
	}
	return best
}

func sortPop(pop []Candidate) {
	// Simple insertion sort (no imports needed).
	for i := 1; i < len(pop); i++ {
		j := i
		for j > 0 && pop[j].Fitness > pop[j-1].Fitness {
			pop[j], pop[j-1] = pop[j-1], pop[j]
			j--
		}
	}
}

// FormatProgram returns a serializable representation of an expression.
func FormatProgram(expr Expr) string {
	return expr.String()
}

// ParseTarget extracts a numeric target from a Forth-style output string like "55 ".
func ParseTarget(s string) (int64, bool) {
	s = strings.TrimSpace(s)
	n, err := fmt.Sscanf(s, "%d", new(int64))
	if n == 1 && err == nil {
		var val int64
		fmt.Sscanf(s, "%d", &val)
		return val, true
	}
	return 0, false
}

File: polyglot/python/evolve.py
"""GP engine using Python AST nodes for symbolic regression."""

import ast
import random
import operator

OPS = {
    ast.Add: operator.add,
    ast.Sub: operator.sub,
    ast.Mult: operator.mul,
    ast.Mod: lambda a, b: a % b if b != 0 else 0,
}
OP_NODES = [ast.Add, ast.Sub, ast.Mult, ast.Mod]


def random_expr(rng, max_depth=4):
    """Generate a random AST expression tree."""
    if max_depth <= 1 or rng.random() < 0.3:
        return ast.Constant(value=rng.randint(0, 100))
    op = rng.choice(OP_NODES)()
    return ast.BinOp(
        left=random_expr(rng, max_depth - 1),
        op=op,
        right=random_expr(rng, max_depth - 1),
    )


def eval_expr(node):
    """Safely evaluate an AST expression tree."""
    if isinstance(node, ast.Constant):
        return node.value
    if isinstance(node, ast.BinOp):
        left = eval_expr(node.left)
        right = eval_expr(node.right)
        fn = OPS.get(type(node.op))
        if fn is None:
            return 0
        try:
            result = fn(left, right)
            if abs(result) > 10**9:
                return 0
            return result
        except Exception:
            return 0
    return 0


def tree_size(node):
    """Count nodes in the tree."""
    if isinstance(node, ast.Constant):
        return 1
    if isinstance(node, ast.BinOp):
        return 1 + tree_size(node.left) + tree_size(node.right)
    return 1


def mutate(rng, node):
    """Return a mutated copy of the expression tree."""
    node = _deep_copy(node)
    choice = rng.randint(0, 2)
    if choice == 0:
        _mutate_constant(rng, node)
    elif choice == 1:
        _mutate_op(rng, node)
    else:
        return _replace_random(rng, node, random_expr(rng, 2))
    return node


def _mutate_constant(rng, node):
    if isinstance(node, ast.Constant):
        node.value += rng.randint(-10, 10)
    elif isinstance(node, ast.BinOp):
        if rng.random() < 0.5:
            _mutate_constant(rng, node.left)
        else:
            _mutate_constant(rng, node.right)


def _mutate_op(rng, node):
    if isinstance(node, ast.BinOp):
        if rng.random() < 0.5 and isinstance(node.left, ast.BinOp):
            _mutate_op(rng, node.left)
        elif isinstance(node.right, ast.BinOp):
            _mutate_op(rng, node.right)
        else:
            node.op = rng.choice(OP_NODES)()


def _replace_random(rng, node, replacement):
    if rng.random() < 0.3:
        return replacement
    if isinstance(node, ast.BinOp):
        if rng.random() < 0.5:
            node.left = _replace_random(rng, node.left, replacement)
        else:
            node.right = _replace_random(rng, node.right, replacement)
    return node


def crossover(rng, a, b):
    """Combine two expression trees."""
    a = _deep_copy(a)
    b = _deep_copy(b)
    donor = _pick_random(rng, b)
    return _replace_random(rng, a, donor)


def _pick_random(rng, node):
    if isinstance(node, ast.BinOp) and rng.random() > 0.3:
        if rng.random() < 0.5:
            return _pick_random(rng, node.left)
        return _pick_random(rng, node.right)
    return _deep_copy(node)


def _deep_copy(node):
    if isinstance(node, ast.Constant):
        return ast.Constant(value=node.value)
    if isinstance(node, ast.BinOp):
        return ast.BinOp(
            left=_deep_copy(node.left),
            op=type(node.op)(),
            right=_deep_copy(node.right),
        )
    return node


def score(expr, target_val):
    """Score a candidate against a target value."""
    result = eval_expr(expr)
    if result == target_val:
        return 1000.0 - tree_size(expr) * 10.0
    diff = abs(result - target_val)
    if diff > 1000:
        return 0.0
    return 1.0 / (1.0 + diff)


def format_program(expr):
    """Serialize expression to a string."""
    if isinstance(expr, ast.Constant):
        return str(expr.value)
    if isinstance(expr, ast.BinOp):
        op_str = {ast.Add: '+', ast.Sub: '-', ast.Mult: '*', ast.Mod: '%'}.get(type(expr.op), '?')
        return f'({op_str} {format_program(expr.left)} {format_program(expr.right)})'
    return '0'


def run_evolution(target_val, max_gens=200, seed=42):
    """Run GP evolution. Returns (best_expr, found_exact)."""
    rng = random.Random(seed)
    pop_size, elites = 30, 5
    pop = [{'expr': random_expr(rng, 4), 'fitness': 0.0} for _ in range(pop_size)]

    for _ in range(max_gens):
        for c in pop:
            c['fitness'] = score(c['expr'], target_val)
        pop.sort(key=lambda c: c['fitness'], reverse=True)
        if pop[0]['fitness'] >= 900.0:
            return pop[0]['expr'], True

        nxt = [{'expr': _deep_copy(c['expr']), 'fitness': 0.0} for c in pop[:elites]]
        while len(nxt) < pop_size:
            parent = _tournament(rng, pop)
            if rng.random() < 0.2:
                other = _tournament(rng, pop)
                child = crossover(rng, parent['expr'], other['expr'])
            else:
                child = mutate(rng, parent['expr'])
            if tree_size(child) > 20:
                child = random_expr(rng, 3)
            nxt.append({'expr': child, 'fitness': 0.0})
        pop = nxt

    for c in pop:
        c['fitness'] = score(c['expr'], target_val)
    pop.sort(key=lambda c: c['fitness'], reverse=True)
    return pop[0]['expr'], pop[0]['fitness'] >= 900.0


def _tournament(rng, pop):
    best = rng.choice(pop)
    for _ in range(2):
        c = rng.choice(pop)
        if c['fitness'] > best['fitness']:
            best = c
    return best


def parse_target(s):
    """Parse a Forth-style target output like '55 '."""
    try:
        return int(s.strip()), True
    except ValueError:
        return 0, False
